adapters/insticator/insticator.go

```diff
@@ -0,0 +1,335 @@
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	if response.Cur != "" {
+		bidResponse.Currency = response.Cur
+	}
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, string, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := json.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	// check if the adUnitId is not empty
+	if insticatorExt.AdUnitId == "" {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: "Missing adUnitId",
+		}
+	}
+
+	// check if the publisherId is not empty
+	if insticatorExt.PublisherId == "" {
+		return openrtb2.Imp{}, insticatorExt.AdUnitId, &errortypes.BadInput{
+			Message: "Missing publisherId",
+		}
+	}
+
+	// Directly construct the impExt
+	impExt := ext{
+		Insticator: impInsticatorExt{
+			AdUnitId:    insticatorExt.AdUnitId,
+			PublisherId: insticatorExt.PublisherId,
+		},
+	}
+
+	impExtJSON, err := json.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		if err := validateVideoParams(imp.Video); err != nil {
+			return openrtb2.Imp{}, insticatorExt.AdUnitId, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+
+	// Return the imp, AdUnitId, and no error
+	return imp, insticatorExt.AdUnitId, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]insticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, caller)
+
+	return json.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return roundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+// roundTo4Decimals function
+func roundTo4Decimals(amount float64) float64 {
+	return math.Round(amount*10000) / 10000
+}
```

Comment:
Consider moving this to our /utils package to help other adapters with the same need.